### PR TITLE
[Flight] Basic Streaming Suspense Support

### PR DESCRIPTION
**Pull request by sebmarkbage**
This adds basic Suspense support to Flight by streaming the response.

First it sends one line with the root JSON model. Within the model holes are replaced with "$" followed by a hex ID. E.g. "$12AB3".

Subsequent lines begin with "J" followed by a hex ID.

So an example response may look something like:

```
{"some":{"json":"$1"},data:["$1","$2"]}
J1:{"more":"data"}
J2:{"even":"more"}
```

The client suspends if you try to access an unresolved value. Once a new line comes in it unsuspends that part of the model.

Currently this suspending is implemented as a getter but it could also be a Proxy. However, more likely is that we'll hide this behind an explicit API that unwraps both components and its data in the future.

## Errors

If the server errors while rendering one of the models that error is encoded with an "E" prefix:

```
{"bad":"$1","good":"$2"}
E1:{"message":"An error happened","stack": "\n..."}
J2:{"this":"worked"}
```

When this happens, the previously suspended part of the model now resolves and starts throwing an error on the client. This will then be caught by an error boundary on the client. Unrelated parts of the response can still continue rendering though.

If the connection dies, then all remaining pending promises resolve and start throwing an error which are caught by error boundaries. Previously resolved parts don't error though.

There's a prioritization scheme so errors are emitted at lower priority in the stream than models.

## Future Additions

I added a comment describing the protocol. The protocol also includes space for a few other types of data.

- _URL_: This can be used to emit URL rows at higher priority than models and can be used to preload other resources without waiting for or even parsing the JSON models yet.
- _Blob_: Traditionally sending binary data is a PITA in existing protocol. We should just have this built-in. It's especially efficient with modern stream APIs. This is important in data heavy use cases like data viz.
- _HTML_: Instead of encoding HTML inside JSON we can just send it as raw HTML for faster parsing and more efficient size. We can also encode that we generated this HTML (using React SSR with proper data structures) in the encoding. So we can by-pass dangerouslySetInnerHTML and just provide a React element that can be rendered directly.

### CODE REVIEW

File: packages/react-server/src/ReactFlightServer.js [hunk 1]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
+    // This is a host element. E.g. HTML.
+    return renderHostChildrenToString(element);
+  } else {
+    throw new Error('Unsupported type.');
+  }
+}
+
+function pingSegment(request: OpaqueRequest, segment: Segment): void {
+  let pingedSegments = request.pingedSegments;
+  pingedSegments.push(segment);
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: Could this segment already be in the array? What would happen in that case? Is it okay to have it twice?

File: packages/react-server/src/ReactFlightServer.js [hunk 2]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
+function retrySegment(request: OpaqueRequest, segment: Segment): void {
+  let value = segment.model;
+  try {
+    while (
+      typeof value === 'object' &&
+      value !== null &&
+      value.$$typeof === REACT_ELEMENT_TYPE
+    ) {
+      // If this is a nested model, there's no need to create another chunk,
+      // we can reuse the existing one and try again.
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: Do I understand correctly this is just an optimization to avoid useless chunks like `J2:"$5"`? It has no impact on correctness?

File: packages/react-server/src/ReactFlightServer.js [hunk 3]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
+    let json = stringify(value, request.toJSON);
+    let row;
+    let id = segment.id;
+    if (id === 0) {
+      row = json + '\n';
+    } else {
+      row = serializeRowHeader('J', id) + json + '\n';
+    }
+    request.completedJSONChunks.push(convertStringToBuffer(row));
+  } catch (x) {
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: Nit: any downside to leaving try-catch around user code more closely? It's weird that making a typo here (e.g. `request.competedJSONChunks`) would lead to streaming an error chunk rather than a crash in the server. Although I guess we have similar undefined behavior in the reconciler code too.

File: packages/react-server/src/ReactFlightServer.js [hunk 4]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
+    if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
+      // Something suspended again, let's pick it back up later.
+      let ping = segment.ping;
+      x.then(ping, ping);
+      return;
+    } else {
+      // This errored, we need to serialize this error to the
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: to the...?

File: packages/react-server/src/ReactFlightServer.js [hunk 5]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
 function performWork(request: OpaqueRequest): void {
-  let rootModel = request.model;
-  request.model = null;
-  let json = JSON.stringify(rootModel, resolveModelToJSON);
-  request.completedChunks.push(convertStringToBuffer(json));
+  let pingedSegments = request.pingedSegments;
+  request.pingedSegments = [];
+  for (let i = 0; i < pingedSegments.length; i++) {
+    let segment = pingedSegments[i];
+    retrySegment(request, segment);
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: Is this guaranteed to not be reentrant? What if we're dealing with a synchronous thenable? Do we assume pings always happen on the next tick?

File: packages/react-server/src/ReactFlightServer.js [hunk 6]
```diff
@@ -42,77 +44,241 @@ type ReactModelObject = {
...
-    for (let i = 0; i < chunks.length; i++) {
-      let chunk = chunks[i];
-      writeChunk(destination, chunk);
+    let jsonChunks = request.completedJSONChunks;
+    let i = 0;
+    for (; i < jsonChunks.length; i++) {
+      request.pendingChunks--;
+      let chunk = jsonChunks[i];
+      if (!writeChunk(destination, chunk)) {
+        request.flowing = false;
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: I don't understand what this means and what the `flowing` property is doing. Can you explain a bit?

File: packages/react-server/src/ReactFlightServer.js [hunk 7]
```diff
@@ -235,7 +235,12 @@ function performWork(request: OpaqueRequest): void {
   }
 }
 
+let reentrant = false;
 function flushCompletedChunks(request: OpaqueRequest): void {
+  if (reentrant) {
+    return;
+  }
+  reentrant = true;
```
Comment (gaearon) [inline packages/react-server/src/ReactFlightServer.js]: I don't understand what this is for.

**Comment by gaearon** [comment]
If connection dies, how do I implement a Retry button?

File: packages/react-flight/src/ReactFlightClient.js [hunk 8]
```diff
@@ -20,100 +20,269 @@ export type ReactModelRoot<T> = {|
...
-  let model = JSON.parse(response.partialRow);
-  Object.defineProperty(modelRoot, 'model', {
-    value: model,
-  });
-  response.rootPing();
+  // In case there are any remaining unresolved chunks, they won't
+  // be resolved now. So we need to issue an error to those.
+  // Ideally we should be able to early bail out if we kept a
+  // ref count of pending chunks.
+  reportGlobalError(response, new Error('Connection closed.'));
```
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-flight/src/ReactFlightClient.js]: We should probably implement the optimization where if all rows are satisfied we don’t even iterate over the set. In that case it’s not as confusing.

File: packages/react-server/src/ReactFlightServer.js (hunk 7 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: I had to hack it because I didn’t realize whatwg streams are reentrant. Will need a better refactor to support both types of a streams.

File: packages/react-server/src/ReactFlightServer.js (hunk 1 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: It’s ok to have it twice but it shouldn’t happen since a component only throws one promise at a time.

File: packages/react-server/src/ReactFlightServer.js (hunk 2 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: Yea barring any bugs.

File: packages/react-server/src/ReactFlightServer.js (hunk 3 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: toJSON can be invoked on objects here which is user code. It’s primarily meant to catch such things. Not implementation bugs.

File: packages/react-server/src/ReactFlightServer.js (hunk 4 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: thing

File: packages/react-server/src/ReactFlightServer.js (hunk 5 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: This whole function is invoked by the “scheduler” (setImmediate) so it’s never reentrant.

Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: The primary purpose of the scheduler is to let multiple things get pinged and flushing to get in before we continue doing work in a batch.

File: packages/react-server/src/ReactFlightServer.js (hunk 6 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js]: Flowing means that we can keep writing more data to the buffer once we have it.

We don’t write eagerly to the output stream if it is full. That lets us emit higher pri things if they become available before the stream buffer becomes not full.

**Comment by sebmarkbage** [comment]
Error boundary and do the request from the root again.

**Comment by gaearon** [comment]
Let me rephrase the tests question. Do you want _me_ to write tests? 😄

**Comment by gaearon** [comment]
I can also just make an app..

**Comment by sebmarkbage** [comment]
If you want to write tests that would be great. Like the suspending and error cases.

**Comment by gaearon** [comment]
Cool, I'll do that.
